Packs/Base/Scripts/CommonServerPython/CommonServerPython.py

```diff
@@ -10865,15 +10868,34 @@ def events_error_handler(res):
         ).format(xsiam_url=xsiam_url, headers=json.dumps(headers, indent=8), status_code=res.status_code, error=error)
 
         demisto.error(header_msg + api_call_info)
-        raise DemistoException(header_msg + error, DemistoException)
+        raise DemistoException(header_msg + error, DemistoException, res)
 
-    zipped_data = gzip.compress(data.encode('utf-8'))   # type: ignore[AttributeError,attr-defined]
+    zipped_data = gzip.compress(data.encode('utf-8'))  # type: ignore[AttributeError,attr-defined]
     client = BaseClient(base_url=xsiam_url)
-    res = client._http_request(method='POST', full_url=urljoin(xsiam_url, '/logs/v1/xsiam'), data=zipped_data,
-                               headers=headers,
-                               error_handler=events_error_handler)
-    if res.get('error').lower() != 'false':
-        raise DemistoException(header_msg + res.get('error'))
+
+    # retry mechanism in case there is a rate limit (429) from xsiam.
+    for attempt_num in range(1, num_of_attempts + 1):
+        demisto.debug('Sending events into xsiam: attempt number {attempt_num}'.format(attempt_num=attempt_num))
+        ok_codes = (200, 429) if attempt_num < num_of_attempts else ()
+        response = client._http_request(
+            method='POST',
+            full_url=urljoin(xsiam_url, '/logs/v1/xsiam'),
+            data=zipped_data,
+            headers=headers,
+            error_handler=events_error_handler,
+            ok_codes=ok_codes,
+            resp_type='response'
+        )
+        demisto.debug('received status code: {status_code}'.format(status_code=response.status_code))
+        if response.status_code == 200:
```

Comment:
best not to use break. Flip the loop to a do while loop with this being one of the conditions

Reply:
changed the implementation to your suggestion :)